### PR TITLE
shader/execution: Port evaluation order tests to flow_control

### PR DESCRIPTION
**Pull request by ben-clayton**
The `flow_control` framework helps with test readability, the errors are easier to understand, and we can add test case permutations, like to allow / disallow constant folding.

This is not an exact porting - some tests have fewer permutations, but I've tried to cover a wider range of interesting cases, like compound assignments and increments.

Issue: #2312

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/execution/flow_control/eval_order.spec.ts [hunk 1]
```diff
@@ -0,0 +1,940 @@
...
+}
+fn b() -> i32 {
+  ${f.expect_order(2)}
+  return 1;
+}`,
+    }));
+  });
+
+g.test('matrix_index_via_ptr')
+  .desc('Test that array indices are evaluated in order, when used via pointers')
```
Comment (amaiorano) [inline src/webgpu/shader/execution/flow_control/eval_order.spec.ts:330]: s/array indices/matrix indices

File: src/webgpu/shader/execution/flow_control/eval_order.spec.ts [hunk 2]
```diff
@@ -0,0 +1,940 @@
...
+fn c() -> f32 {
+  ${f.expect_order(3)}
+  return 3;
+}
+`,
+    }));
+  });
+
+g.test('nested_builtin_fn_args')
+  .desc('Test builtin function call arguments are evaluated left-to-right')
```
Comment (amaiorano) [inline src/webgpu/shader/execution/flow_control/eval_order.spec.ts:564]: Test "nested" builtin function call arguments...

**Review by amaiorano — APPROVED** [review]
I love it!

File: src/webgpu/shader/execution/flow_control/eval_order.spec.ts (hunk 1 above)
Comment (ben-clayton) [reply src/webgpu/shader/execution/flow_control/eval_order.spec.ts]: Done

File: src/webgpu/shader/execution/flow_control/eval_order.spec.ts (hunk 2 above)
Comment (ben-clayton) [reply src/webgpu/shader/execution/flow_control/eval_order.spec.ts:564]: Done

**Review by ben-clayton — COMMENTED** [review]
Thank you!

**Comment by ben-clayton** [comment]
(helps to save before pushing) Really done.

**Pull request merged**